src/models/transformer.py

```diff
@@ -1,29 +1,37 @@
 import torch
-from transformers import GPT2Config, GPT2Model # type: ignore
+from transformers import GPT2Config, GPT2Model, LlamaConfig, LlamaModel
 from torch import nn
 
 from core import ContextModel
 
-
+ARCHITECTURES = {
+    "gpt2": {'config': GPT2Config, 'model': GPT2Model},
+    "llama2": {'config': LlamaConfig, 'model': LlamaModel}
+} 
 class TransformerModel(ContextModel):
     def __init__(self, x_dim, n_positions, n_embd=128, n_layer=12, n_head=4, **kwargs):
         super(TransformerModel, self).__init__()
-        configuration = GPT2Config(
+        architecture = ARCHITECTURES[kwargs["architecture"]]
+        configuration = architecture['config'](
             n_positions=2 * n_positions,
+            max_position_embeddings=2 * n_positions,
             n_embd=n_embd,
+            hidden_size=n_embd,
             n_layer=n_layer,
+            num_hidden_layers=n_layer,
             n_head=n_head,
+            num_attention_heads=n_head,
             resid_pdrop=0.0,
             embd_pdrop=0.0,
             attn_pdrop=0.0,
             use_cache=False,
         )
-        self.name = f"gpt2_embd={n_embd}_layer={n_layer}_head={n_head}"
+        self.name = f"{kwargs['architecture']}_embd={n_embd}_layer={n_layer}_head={n_head}"
 
         self.context_length = n_positions
         self._n_dims = x_dim
         self._read_in = nn.Linear(x_dim, n_embd)
-        self._backbone = GPT2Model(configuration)
+        self._backbone = architecture['model'](configuration)
```

Comment:
I like the code reuse here, could we take a more delegation-esque approach? For example, have a base class that takes in just the backbone method and have two very thin classes that serve for being either GPT2 or Llama2

Reply:
implemented